config/initializers/i18n.rb

```diff
@@ -45,3 +45,35 @@ def store_translations(locale, data, options = {})
 
   I18n.available_locales
 end
+
+AVAILABLE_LANGUAGES = Rails.root.glob("config/locales/*.yml").map do |filename|
+  code = File.basename(filename, ".yml")
+  native_name = case code
+                when "en"
+                  "English"
+                when "en-GB"
+                  "English (United Kingdom)"
+                else
+                  line = File.open(filename, &:readline)
+                  line.match(/\(([^(]*\([^(]*\))\)$/) do |match_data|
+                    match_data[1]
+                  end || line.match(/\(([^(]*)\)$/) do |match_data|
+                    match_data[1]
+                  end || code
+                end
+
+  {
+    :code => code,
+    :native_name => native_name
+  }
+end
+
+AVAILABLE_LANGUAGES.sort_by! do |entry|
+  # https://stackoverflow.com/a/74029319
+  diactrics = [*0x1DC0..0x1DFF, *0x0300..0x036F, *0xFE20..0xFE2F].pack("U*")
+  entry[:native_name]
+    .downcase
+    .unicode_normalize(:nfd)
+    .tr(diactrics, "")
+    .unicode_normalize(:nfc)
+end
```

Comment:
The more correct way to do this is to use a gem like `ffi-icu` or `twitter_cldr` that can do proper unicode sorting, for example with `ffi-icu` you would do something like:

```ruby
collator = ICU::Collation::Collator.new("root")
collator.strength = :primary
AVAILABLE_LANGUAGES.sort! do |a, b|
  collator.compare(a, b)
end
```
Reducing strength from the default of `:tertiary` to `:primary` makes the sort case and accent insensitive though even at tertiary it still sorts by the base character first and only uses accent (at level secondary) and then case (at level tertiary) to break ties which probably makes little practical difference for the data set here.

Reply:
Isn't it then better to dump all languages to some yaml file in a predetermined order, rather than adding another gem?

Reply:
dumped to `config/ui_languages.yml`